stage3.0_first_views/traited_face_detect.py

```diff
@@ -0,0 +1,116 @@
+# General imports
+from os.path import join
+
+import PIL.Image
+import numpy as np
+from matplotlib import patches
+from matplotlib.figure import Figure
+from PIL.ExifTags import TAGS
+from skimage import data
+from skimage.feature import Cascade
+
+# ETS imports
+from traits.api import (
+    Dict,
+    File,
+    HasStrictTraits,
+    Instance,
+    List,
+    observe,
+)
+from traitsui.api import Item, ModelView, View
+
+# Local imports
+from ets_tutorial.util.mpl_figure_editor import MplFigureEditor
+
+
+class ImageFile(HasStrictTraits):
+    """ Model to hold an image file.
+    """
+    filepath = File
+
+    faces = List
+
+    metadata = Dict
+
+    def to_array(self):
+        if not self.filepath:
+            return np.array([])
+
+        with PIL.Image.open(self.filepath) as img:
+            return np.asarray(img)
+
+    @observe("filepath")
+    def _update_faces_and_metadata(self, event):
+        self.metadata = {}
+        self._update_metadata_with_exif()
+        self._detect_faces()
+
+    def _update_metadata_with_exif(self):
+        if not self.filepath:
+            return
+        with PIL.Image.open(self.filepath) as img:
+            exif = img._getexif()
+        if not exif:
+            return
+        self.metadata.update(
+            {TAGS[k]: v for k, v in exif.items() if k in TAGS}
+        )
+
+    def _detect_faces(self):
+        self.faces = []
+        if not self.filepath:
+            return
+        # Load the trained file from the module root.
+        trained_file = data.lbp_frontal_face_cascade_filename()
+        # Initialize the detector cascade.
+        detector = Cascade(trained_file)
+        self.faces = detector.detect_multi_scale(
+            img=self.to_array(),
+            scale_factor=1.2,
+            step_ratio=1,
+            min_size=(60, 60),
+            max_size=(600, 600)
+        )
+        self.metadata['Number of faces'] = len(self.faces)
+
+
+class ImageFileView(ModelView):
+    """ ModelView for an image file object.
+    """
+    model = Instance(ImageFile)
+
+    figure = Instance(Figure)
+
+    view = View(
+        Item("model.filepath", style="readonly", show_label=False),
```

Comment:
Do we want to make this not readonly so people can leverage the `build_mpl_figure` listener below?

Reply:
Or keep it readonly in 3.1 but make it "live" and add the `@observe("model.filepath")` in 3.2?

Reply:
> Do we want to make this not readonly so people can leverage the build_mpl_figure listener below?

Great idea! This needed #44 to function properly so we can go ahead and do this now.

> Or keep it readonly in 3.1 but make it "live" and add the @observe("model.filepath") in 3.2?

Interesting. Let me think a little more on how we could version this segment of the tutorial.

Reply:
Side note: We should track the difference between minor versions (i.e, 3.0, 3.1, 3.2), so that the presenter's job is a little easier. I'll try to add some "developer docs" with this PR.

Reply:
Yes!!!